### DISCUSSION TITLE
SimpleX Chat

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#367] [General]
https://simplex.chat/

This has become a beloved messenger, despite the horrible name, by people on the Nostr network. Jack from Twitter has endorsed it often.

The main thing about it is that the protocol seems to be even more private than sealed sender networks like Signal.

> Many users asked: if SimpleX has no user identifiers, how can it know where to deliver messages?

> To deliver messages, instead of user IDs used by all other platforms, SimpleX uses temporary anonymous pairwise identifiers of message queues, separate for each of your connections — there are no long term identifiers.

> You define which server(s) to use to receive the messages, your contacts — the servers you use to send the messages to them. Every conversation is likely to use two different servers.

> This design prevents leaking any users' metadata on the application level. To further improve privacy and protect your IP address you can connect to messaging servers via Tor.

> Only client devices store user profiles, contacts and groups; the messages are sent with 2-layer End-to-end encryption.

> Read more in [SimpleX whitepaper](https://github.com/simplex-chat/simplexmq/blob/stable/protocol/overview-tjr.md).

Personally, I find the privacy on Reticulum / Sideband to be more than adequate, but one feature that I like of SimpleX is how easy it is to manage multiple identities in the one app. But on Reticulum, it seems like that could also be done on the application level.

**Comment by markqvist** [comment]
That platform comes up every now and then, so I'll just reiterate what I've said earlier in various places. **TL;DR: I'm not impressed, and slightly suspicious.**

The marketing material of SimpleX is on such a high horse about their purported security and privacy. But when you actually look into it, things fall apart.

Their headline marketing:

> The first messaging platform that has no user identifiers of any kind - 100% private by design

This is complete marketing bullshit. In no way was SimpleX "the first" of anything, and only in some fairy universe is anything "100% private by design". Claiming so in just misleading users, and playing on a "we're better than *them*", with "them" being all other platforms.

> To deliver mesages, instead of user IDs used by all other platforms, SimpleX has identifiers for message queues, separate for each of your contacts. In the current version of the protocol each queue is used until the contact is deleted. Later this year we plan to add queue rotation to the client protocol

That is a very roundabout way of admitting that the app does indeed have "user indentifiers", and that they "maybe soon" will do something about it.

Other insane bullshit from their website: [Claiming that other End-to-End Encrypted services are susceptible to Man-in-the-Middle Attacks](https://simplex.chat/#problem-explained), and that they are then somehow the solution to that.

I don't know who came up with that dumpster fire of a marketing strategy, but it makes them totally uncredible in my eyes.

Also, their system still completely relies on *their* servers to be functional and serving users, and their entire infrastructure could be changed in a snap by the developers. It is not decentralised, and therefore users have no actual control.

Just from a cursory reading of their material on GitHub, there is several other things that stick out to me as both potential problems and misinformation about their product.

I would be curious to actually read and review their implementation and system design, and see what things might pop up.

In general, I'd say that users are much better off using apps based on the [Matrix](https://matrix.org) protocol, or Signal (if you really must, but use Matrix if you can).

**Reply by trosel** [reply]
> Also, their system still completely relies on their servers to be functional and serving users, and their entire infrastructure could be changed in a snap by the developers. It is not decentralised, and therefore users have no actual control.

This is also a concern of mine. They are totally subsidizing the cost of their own network with their investors’ money. How do they plan to incentivize the network without that? They will charge for “SimpleX premium”? How will users pay without doxxing themselves?

**Reply by markqvist** [reply]
Yes, very well put. The same types of concerns I have been having. Unfortunately I think it will play out in a way that is not very benificial for the users. Promise everything and ride the hype, but when the bills fall due, someone has to pay.

**Reply by faragher** [reply]
It generates a particularly fault intolerant network as well. If a server is needed, and a known path to that server is also required, then you already have two points that can kill the entire system.

**Comment by markqvist** [comment]
On another note, Sideband could really use an easy way to switch identities though! Completely agree on that point.

**Comment by faragher** [comment]
What Mark said, but with more moving parts:

I won't get into "first" or any of the other marketing, but anyone who tells you something is 100% secure is delusional, ignorant, or selling you a bill of goods. I have found exactly one form of 100% secure cryptography (a truly random one time pad) that is still defeatable (copying the pad, running out of entropy). "Cryptographically secure" is simply reasonable secure for the ecosystem as we know it. That being said, if the modern implementations aren't secure, nations fall, banks empty, civilization goes back 50-100 years *at best.*

So let's just look at their claims, since they're not all bad.

### 1: No user identifiers.

Blatantly false. You need an address to send something. 

What they mean: No user accounts, or central identity.

This is true (as far as I can tell). You make a new numbered account, send a message or two to a known number, then burn it. Neat. This mirrors Reticulum's handling of identities. Use and discard at will (with Reticulum adding a little more utility with each appname/aspect pair having a unique address under an identity)

### 2: Out of Band key exchanges prevent MitM attacks

Misleading, and a potential security hole. Passing a symmetric key around out of band does, in fact, prevent people from stealing it on the wire. But it also adds multiple new ways of intercepting it or even requires face to face contact *every single time the key changes.* These are absurd security flaws and undermines the entire point of rapidly changing keys. In vastly increases the number of messages at risk or requires risky behavior to change keys rapidly.

Asymmetric encryption was designed to remedy this situation. A public key is useless except to send something to a private key, so share it in the clear. Use this quite secure (but computationally intensive) method to pass an ephemeral symmetric key. This works only as long as it needs to, and can be swapped at any interval. No contact is required, and a public key announce means that it's even deniable that two actors have ever interacted. This is how Reticulum implements encryption (although I admit the specifics are still a little beyond me)

### 3: Identities and messages are stored encrypted locally, and servers only hold the E2EE data until it's passed along

Good. This is an excellent way of handling things, and I have no notes. This is also how Reticulum works . . . although I'm unsure if it's encrypted on disk. I know the private identity is vulnerable, but unless you're typing a password on every program, there's always going to be sensitive data files on your machine.

### 4: You can make your own servers or use theirs, and they're independent of Cryptocurrencies.

Beautiful. Seriously, I'm sad this even has to be said these days, since who wants an immutable record blockchain of their "private" communications? I'd have to dig into specifics, but this is exactly how things should be done: start your own network at will. So long as they interoperate with no preference, life is good.

### 5: Encrypted messages, calls, and files

Neat, but data is data. This is obviously built on top of a high speed network, and the addressing is really the only new part of it, but it's what I'd expect from the average consumer product. 

### 6: Disappearing messages

I mean . . . this is application level. I think SnapChat has taught us that faith in an uncompromised end user is a bad idea, but it's something people like seeing. I'd never trust it.

### 7: Incognito mode

I . . . umm . . . I'm sorry, some of your secure communications are okay to leave lying on your phone, but others aren't?

### 8: "Perfect" forward secrecy and defense against traffic correlation

Again, this requires a bit more than "I'll show you a QR code" level of key exchanges. And traffic correlation and time of flight analysis can only be defeated by large amounts of random traffic, combined with some active countermeasures. If there's only six people talking on a network, do you really think there's any real security against knowing who's talking to whom?

### 9: Message authentication, encryption beyond TLS, and uses TLS

This should be expected.

### 10: Access via TOR

Not a bad option. Presupposes connection over IP and contains all the other security issues with TOR.

### 11: Content padding

Yes, that is how block ciphers work.

### 12: No centralized components or dependence on DNS

Well, either you have a list of IP addresses, and this creates centralization, or you rely on DNS to point to multiple servers. Or maybe you rely on somehow contacting the Internet to get new IPs, but that then presupposes the above problem once again.



I like that these systems exist. I will never say that they should stop trying or that a plurality of systems is a bad thing. However, this is another example of slick marketing that gets people to buy into a big idea with lots of buzz words as opposed to something more fundamental. Does Facebook Messenger have its place? Of course! Will Reticulum ever replace POP/SMTP? Nope. Are FTP, IRC, and old systems that are useful but unpopular still in service? Yup. It's good to let these things grow and breathe, and it'll work or it won't. So long as they're not just milking money from people or otherwise preying on them, I'm all for it! 

It's cool to see stuff like this. But I'd absolutely never use it professionally.

**Comment by epoberezkin** [comment]
Hello! Evgeny here. Our users always share when there is some interesting discussion.

There is some mixture of valid criticism, limitations, and some misconceptions too, so I'll comment!

**Comment by epoberezkin** [comment]
> The marketing material of SimpleX is on such a high horse about their purported security and privacy. But when you actually look into it, things fall apart.
> The first messaging platform that has no user identifiers of any kind - 100% private by design
> This is complete marketing bullshit. In no way was SimpleX "the first" of anything, and only in some fairy universe is anything "100% private by design". Claiming so in just misleading users, and playing on a "we're better than them", with "them" being all other platforms.

While it's clear that you dislike our marketing style, it shouldn't detract your attention from factual advantages and disadvantages, and from reading the below it seems like it does.

Most marketing is to some extent exaggeration, and you are saying that we are not allowed to exaggerate at all.

>  In no way was SimpleX "the first" of anything

That depends on what we define as "user identities", and also on the definition of what is "messaging platform". Provided it's correct that there are no user identities (more on that below) and that not every communication solution is a "messaging platform/network", then this claim _might be_ correct, but **please let me know your counter-examples - what messaging platform or network that had no user identities of any kind was there before SimpleX?**

In any case, the tagline on our website is different now, for a long time:

Privacy redefined: The first messenger without user IDs

I can comment on it, if you are interested.

> and only in some fairy universe is anything "100% private by design".

I think it's obvious to everybody, and I don't think there is anybody who thinks otherwise, but it's a plain marketing-speak that is commonly used. SimpleX Chat, with some limitations, is more private than many alternative, but I don't know how to measure it in %s.

> Claiming so in just misleading users,

That I actually strongly disagree with. For something to be misleading there should be either an intention to mislead or somebody has to be actually misled by this statement, and it's purposefully exaggerated exactly to avoid it being misleading.

By saying that this is misleading to the users, you are effectively saying "I am smart enough to see it as marketing bullshit and _users_ are not". I don't know who are these dumb users who can't see 100% as a marketing exaggeration, sorry. I am absolutely positive that 100% of our users who _need_ privacy don't take "100% private by design" as anything other than marketing language it is. We see 100% private as a destination, and something that attracts more helpful criticism and results in faster improvements. At the same time we are aiming to be very factual and specific about the design and its limitations. See these docs for example:

https://github.com/simplex-chat/simplexmq/blob/stable/protocol/overview-tjr.md#threat-model
https://github.com/simplex-chat/simplex-chat?tab=readme-ov-file#privacy-and-security-technical-details-and-limitations

> and playing on a "we're better than them", with "them" being all other platforms.

Comparative marketing is allowed, and if what we do is factually better we are allowed to say it.

E.g., in terms of e2e encryption security SimpleX Chat is factually more secure than most alternatives, at least design-wise - there are obvious caveats of lower maturity, lower adoption, and fewer security reviews (two more are coming this year though - design and implementation). Please see the comparison here: https://simplex.chat/blog/20240314-simplex-chat-v5-6-quantum-resistance-signal-double-ratchet-algorithm.html

> To deliver mesages, instead of user IDs used by all other platforms, SimpleX has identifiers for message queues, separate for each of your contacts. In the current version of the protocol each queue is used until the contact is deleted. Later this year we plan to add queue rotation to the client protocol
> That is a very roundabout way of admitting that the app does indeed have "user indentifiers", and that they "maybe soon" will do something about it.

@faragher
>  1. No user identifiers.
> Blatantly false. You need an address to send something.
> What they mean: No user accounts, or central identity.

That is absolutely not correct. Of course there are no accounts, but **we do mean that there are no _user_ identifiers of any kind**, that is absolutely nothing that uniquely identifies users on the protocol level to their contacts or to the relays. The statement "You need an address to send something" is correct, but "a delivery address" only represents a _user profile identifier_ if it's used for all communication contexts by a given user profile (like it is the case with Session or Cwtch IDs). A temporary delivery address used in SimpleX for a given conversation with one contact only is certainly not a user identifier.

If you imagine communication network as a graph, where users are vertices and connections between users are edges, then vertices in SimpleX network (users) have no identifiers, only the edges do. Connections between users have identities of course, but even they are not permanent - they are temporary, as the conversation can be moved to another set of relays, changing these identifiers. Currently it's manual, and will be automated. The identifiers of these connections are anonymous pairwise identifiers (see NIST dictionary of terms for example), and they do NOT represent identities of the users.

> Other insane bullshit from their website: [Claiming that other End-to-End Encrypted services are susceptible to Man-in-the-Middle Attacks](https://simplex.chat/#problem-explained), and that they are then somehow the solution to that.

I would refer you to the linked post above for more elaboration on MITM, security of key exchange, and possible mitigations of MITM attacks on key exchange by the operator. SimpleX is not the only network that offers _non-optional_ mitigation to MITM attacks by introducing second factor for key exchange - some others do it too. In our case we include keys into the connection invitation, Cwtch includes Tor hidden service address - in any case if the address that exchanged out of band contains some key material, then such key exchange cannot be easily compromised by the operator. Absolute majority of messaging platforms don't require a non-optional second factor to key exchange - at best they offer additional, optional verification of key exchange after the connection is made (that is, **detection rather than prevention of MITM**), and most users don't use it. Even people who did validate security code on Signal after the first connection, usually don't bother doing it again when "security code changed". But unfortunately it is not possible to differentiate between security code changing because of user's contact changing the device or because of MITM via the operator, so without re-validating security code out-of-band every time it changes, it can be compromised.

So the difference you are criticising is between non-optional prevention of MITM attack in SimpleX (and Cwtch, for what it's worth) and optional detection of MITM attack in Signal - to me, this difference is substantial enough. You may have a different opinion about whether the difference between detection and prevention is substantial enough, but it's just factual and technical reality of key exchange security or lack of it, rather than "insane bullshit", as you call it.

@faragher:
>  2. Out of Band key exchanges prevent MitM attacks
> Misleading, and a potential security hole. Passing a symmetric key around out of band does, in fact, prevent people from stealing it on the wire. But it also adds multiple new ways of intercepting it or even requires face to face contact every single time the key changes.

We of course don't pass symmetric keys, we only pass one of the peer's asymmetric keys out of band, passing symmetric keys would be insecure. So there is nothing misleading here, and you need to review the design before making such false claims.

> I don't know who came up with that dumpster fire of a marketing strategy, but it makes them totally uncredible in my eyes.

Totally me to blame, we do what we can... [Esra'a who just joined](https://simplex.chat/blog/20240404-why-i-joined-simplex-chat-esraa-al-shafei.html) will certainly help us to improve here.

> Also, their system still completely relies on their servers to be functional and serving users, and their entire infrastructure could be changed in a snap by the developers. It is not decentralised, and therefore users have no actual control.

This is incorrect. Users choose which servers to use, and there are many 100s if not 1000s self-hosted servers in SimpleX network, and several third party apps also rely on our code for their in-app communications - and they don't rely on our servers, and we can't even know about all of them.

What is true is that we didn't create yet is a model to incentifvize other network operators, but it's certainly coming - we see decentralization of network operators offered via the app as a high priority.

At the same time, the model of decentralisation where all relays are known and can be used by all clients appears to be a less decentralised design (e.g., Tor with it's central authority and centralised list of the servers).

So a trade off here is that we are choosing a slower path to a higher degree of decentralization - if we chose design similar to Tor (or had a hard dependency on Tor) then the initial decentralization would have been higher, but in the end it would be lower than SimpleX design allows. Same about pure p2p designs - it results in higher initial decentralization, but in lower privacy, and in lower eventual decentralization. Hope it makes sense, but happy to elaborate further.

> Just from a cursory reading of their material on GitHub, there is several other things that stick out to me as both potential problems and misinformation about their product.

Without being specific, it's just overly emotional and baseless accusations. While you don't have to be a fan, in your line of business you have to remain professional and be able to substantiate it - please do.

> I would be curious to actually read and review their implementation and system design, and see what things might pop up.

Any substantive criticism would be welcome - we're always looking for what needs to be improved.

> In general, I'd say that users are much better off using apps based on the [Matrix](https://matrix.org/) protocol, or Signal (if you really must, but use Matrix if you can).

Yes, but Matrix encryption remains broken and not protected from MITM attack, and Signal's one-line client code modification exposes "private" phone numbers in some cases, but both are indeed much more mature and much better funded apps. So it's a trade off for the users to make, but without elaborating on individual circumstances of said users, it wrong to say they are better off with Matrix or with Signal (or even with Cwtch, which surprisingly isn't mentioned) - it very much depends on their threat model and requirements.

@faragher
> 3. Identities and messages are stored encrypted locally
> although I'm unsure if it's encrypted on disk. 

Yes, both database and all files are encrypted in storage - each file is encrypted with a different key that is associated with the message, so removing messages in the database makes files not decryptable even if their encrypted copy is retained and database is later compromised (an equivalent of forward secrecy for local storage).

> 7: Incognito mode
> I . . . umm . . . I'm sorry, some of your secure communications are okay to leave lying on your phone, but others aren't?

Not sure what is the criticism here, but I would direct you to this post explaining what incognito mode is:

https://simplex.chat/blog/20220901-simplex-chat-v3.2-incognito-mode.html#incognito-mode

> 11: Content padding
> Yes, that is how block ciphers work.

Block ciphers pad to a block size, making correlation by message size still very easy. Very few messengers pad all messages to fixed packet size. SimpleX pads to 16kb, Cwtch to 8kb, I don't know what other messenger (since Pond) does it. Padding schemes like Padme or just padding to a small encryption block doesn't solve the problem of traffic correlation by message size, padding to a fixed block size does.

> 12. No centralized components or dependence on DNS
> Well, either you have a list of IP addresses, and this creates centralization, or you rely on DNS to point to multiple servers. Or maybe you rely on somehow contacting the Internet to get new IPs, but that then presupposes the above problem once again.

DNS may be used for addressing the relays, but it doesn't have to be - so it's not a hard dependency. Lots of users just host relays on .onion addresses, to create subnetworks accessible only via Tor, and some just host them on IP addresses, without any DNS.

So no, there is no lists of IP. And there is no hard dependency on DNS. And no, clients don't depend on any source of the servers. Clients have to decide which servers to use, and currently there are preset servers we operate, as this is, effectively, a growing prototype network. As I wrote, the objective is to have multiple providers offered via the app, and any other providers configurable, and by default every single conversation will be happening via 4 different providers (4 relays, each chosen from a different provider - currently it's 2 relays, but will be 4 within 2-3 months).

--

The largest point of criticism that must have been mentioned, but wasn't, is the limitations of IP address protection - as relays are chosen by message recipients, they can modify their code to monitor IP addresses of connecting users. It's mentioned on site home page, and in many other places, so it takes users about 10 minutes to discover, and they either accept this risk or use Tor or VPN.

This is by far the largest current limitation of SimpleX network, and the most frequently demanded by the users - so it is coming it 2-3 months (the switch from 2 to 4 relays per conversation is exactly about this limitation).

So with all the incorrect statements, it's somewhat ironic that the strongest critic is absent... Anyway, I was happy to fill this gap.

--

To summarise, I got your message about the need to make our marketing more professional, I agree, and we see it as a priority as well.

I think that being engineers you could have had a bit more patience and capacity to analyse things deeper, or at least ask questions, and not simply to jump to incorrect conclusions, or make some wild accusations not based on facts - it's not very professional.

And I have no issue engaging in comparative and factual criticism, but I don't think that misleading the users about how things work, as some of your statements do, is the right thing to do.

It does seem like it comes from the misunderstanding of SimpleX design, and, maybe, seeing the conspiracy and bad intentions where there isn't any, rather than from your own bad intentions, so no hard feelings here.

Please just remember that we're all human, and we just do the best we can. I am happy to answer any specific questions and provide specific clarifications on any aspects of our design or implementation.

I'll also look deeper into Reticulum, and hopefully it's better than many things, and I wish you it grows as fast as SimpleX Chat does - we all need more tools.

**Reply by faragher** [reply]
Thank you for your time and clarifications. By changing your marketing and feature list to be more technical with more specifics will certainly be helpful. I had only the statements on record at the time to judge your system, I don't have time to reverse engineer a project to determine the function, and thus all I have to go on is your publications.

The multiple small nit-picks are because the marketing speak and claims were consistently over-reaching. As you're addressing this problem, I have little issue. Seeing a problem, taking moves to correct it, and addressing specifics when required is the correct way to go.

I do wish your project the best: a rising tide floats all boats and we're both obviously interested in the same underlying needs and issues. The fact that you took your time to specifically come down here and explain where things were unclear is appreciated.
